core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -0,0 +1,312 @@
+    // /multipart/src/main/java/io/helidon/media/multipart/MultiPartDecoder.java
+
+    private static final AtomicReferenceFieldUpdater<MultipartDecoder, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(
+                    MultipartDecoder.class, Subscription.class, "upstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<MultipartDecoder, Subscriber> downstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(
+                    MultipartDecoder.class, Subscriber.class, "downstream");
+
+    private final CompletableFuture<BufferedEmittingPublisher<BodyPart>> initFuture;
+    private final LinkedList<BodyPart> bodyParts;
+    private final MimeParser parser;
+    private final ParserEventProcessor parserEventProcessor;
+
+    @Nullable
+    private BufferedEmittingPublisher<BodyPart> emitter;
+    @Nullable
+    private BodyPartBuilder bodyPartBuilder;
+    @Nullable
+    private HttpHeadersBuilder bodyPartHeaderBuilder;
+    @Nullable
+    private BufferedEmittingPublisher<HttpData> bodyPartPublisher;
+
+    // Updated only via upstreamUpdater
+    @Nullable
+    private volatile Subscription upstream;
+    // Updated only via downstreamUpdater
+    @Nullable
+    private volatile Subscriber<? super BodyPart> downstream;
+
+    MultipartDecoder(String boundary) {
+        requireNonNull(boundary, "boundary");
+        parserEventProcessor = new ParserEventProcessor();
+        parser = new MimeParser(boundary, parserEventProcessor);
+        initFuture = new CompletableFuture<>();
+        bodyParts = new LinkedList<>();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super BodyPart> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        if (emitter != null || !downstreamUpdater.compareAndSet(this, null, subscriber)) {
+            subscriber.onSubscribe(SubscriptionHelper.CANCELLED);
+            subscriber.onError(new IllegalStateException("Only one Subscriber allowed"));
+            return;
+        }
+        deferredInit();
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (!upstreamUpdater.compareAndSet(this, null, subscription)) {
+            subscription.cancel();
+            throw new IllegalStateException("Subscription already set.");
+        }
+        deferredInit();
+    }
+
+    @Override
+    public void onNext(HttpData chunk) {
+        try {
+            final ByteBuf byteBuf;
+            if (chunk instanceof ByteBufHolder) {
+                byteBuf = ((ByteBufHolder) chunk).content();
+            } else {
+                byteBuf = Unpooled.wrappedBuffer(chunk.array());
+            }
+            parser.offer(byteBuf);
+            parser.parse();
+        } catch (MimeParsingException ex) {
+            emitter.fail(ex);
+            ReferenceCountUtil.release(chunk);
```

Comment:
Should we cancel the subscription and return here?

Reply:
That's a good idea.